cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,681 @@
+        cugraph::edge_dst_dummy_property_t{}.view(),
+        cugraph::edge_dummy_property_t{}.view(),
+        [] __device__(auto src, auto dst, thrust::nullopt_t, thrust::nullopt_t, thrust::nullopt_t) {
+          return true;
+        },
+        edge_mask.mutable_view(),
+        false);
+
+      cur_graph_view.attach_edge_mask(edge_mask.view());
+
+      // case 3. For the (p, q), find intersection 'r' to create (p, r, -1) and (q, r, -1)
+      // FIXME: check if 'invalid_edge_first' is necessery as I operate on 'vertex_pair_buffer'
+      // which contains the ordering with the number of triangles.
+      // FIXME: debug this stage. There are edges that have been removed that are still found in nbr
+      // intersection
+      auto [intersection_offsets, intersection_indices] =
+        detail::nbr_intersection(handle,
+                                 cur_graph_view,
+                                 cugraph::edge_dummy_property_t{}.view(),
+                                 get_dataframe_buffer_begin(invalid_edges_buffer),
+                                 get_dataframe_buffer_end(invalid_edges_buffer),
+                                 std::array<bool, 2>{true, true},
+                                 do_expensive_check);
+
+      size_t accumulate_pair_size = intersection_indices.size();
+
+      auto vertex_pair_buffer_p_r_edge_p_q =
+        allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(accumulate_pair_size,
+                                                                     handle.get_stream());
+
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+        get_dataframe_buffer_end(vertex_pair_buffer_p_r_edge_p_q),
+        generate_p_r_q_r<vertex_t,
+                         edge_t,
+                         decltype(get_dataframe_buffer_begin(invalid_edges_buffer))>{
+          0,
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          get_dataframe_buffer_begin(invalid_edges_buffer)});
+
+      auto edge_last       = transposed_edge_first + edgelist_srcs.size();
+      auto num_edge_exists = accumulate_pair_size;
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(num_edge_exists),
+                       unroll_edge<vertex_t, edge_t, decltype(transposed_edge_first)>{
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+                         transposed_edge_first,
+                         edge_last,
+                       });
```

Comment:
This code for unrolling (p, r) contribution and the following code unrolling (q,r) contribution is nearly identical. This code should be refactored to avoid  code repetition.

Reply:
Right.